zcash_client_sqlite/src/wallet.rs

```diff
@@ -519,28 +521,264 @@ pub(crate) fn get_balance(
+        block_fully_scanned(conn)?.map_or(birthday_height - 1, |m| m.block_height());
+
+    let sapling_scan_progress =
+        sapling_scan_progress(conn, fully_scanned_height, chain_tip_height)?;
+
+    let summary_height = chain_tip_height + 1 - min_confirmations;
+
+    // If any of the shard below the anchor (or chain tip if no anchor is specified) is unscanned
+    // above the wallet birthday height, none of our balance is currently spendable.
+    let any_spendable = conn
+        .query_row(
+            "SELECT 1 FROM v_sapling_shard_unscanned_ranges
+             WHERE :summary_height
+                BETWEEN subtree_start_height
+                AND IFNULL(subtree_end_height, :summary_height)
+             AND block_range_start <= :summary_height",
+            named_params![":summary_height": u32::from(summary_height)],
+            |_| Ok(true),
+        )
+        .optional()?
+        .is_none();
+
+    let mut stmt_select_notes = conn.prepare_cached(
+        "SELECT n.account, n.value, n.is_change, scan_state.max_priority, t.block, t.expiry_height
+        FROM sapling_received_notes n
+        LEFT OUTER JOIN transactions t ON t.id_tx = n.tx
+        INNER JOIN v_sapling_shards_scan_state scan_state
+            ON  n.commitment_tree_position >= scan_state.start_position
+            AND n.commitment_tree_position <  scan_state.end_position_exclusive
+        WHERE n.spent IS NULL",
     )?;
 
-    match Amount::from_i64(balance) {
-        Ok(amount) if !amount.is_negative() => Ok(amount),
-        _ => Err(SqliteClientError::CorruptedData(
-            "Sum of values in sapling_received_notes is out of range".to_string(),
-        )),
+    let mut account_balances: BTreeMap<AccountId, AccountBalance> = BTreeMap::new();
+    let mut rows = stmt_select_notes.query([])?;
+    while let Some(row) = rows.next()? {
+        let account = row.get::<_, u32>(0).map(AccountId::from)?;
+
+        let value_raw = row.get::<_, i64>(1)?;
+        let value = NonNegativeAmount::from_nonnegative_i64(value_raw).map_err(|_| {
+            SqliteClientError::CorruptedData(format!("Negative received note value: {}", value_raw))
+        })?;
+
+        let is_change = row.get::<_, bool>(2)?;
+
+        let max_priority_raw = row.get::<_, i64>(3)?;
+        let max_priority = parse_priority_code(max_priority_raw).ok_or_else(|| {
+            SqliteClientError::CorruptedData(format!(
+                "Priority code {} not recognized.",
+                max_priority_raw
+            ))
+        })?;
+
+        let received_height = row
+            .get::<_, Option<u32>>(4)
+            .map(|opt| opt.map(BlockHeight::from))?;
+
+        let mut spendable_value = NonNegativeAmount::ZERO;
+        let mut change_pending_confirmation = NonNegativeAmount::ZERO;
+        let mut value_pending_spendability = NonNegativeAmount::ZERO;
+
+        let is_spendable = any_spendable
+            && received_height.iter().any(|h| h <= &summary_height)
```

Comment:
I continue to prefer `received_height.map_or(false, |h| h <= &summary_height)`, but I won't press the issue.